iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/cache/CacheManager.java

```diff
@@ -287,17 +287,26 @@ public List<ICachedMNode> collectVolatileMNodes() {
   }
 
   private void collectVolatileNodes(ICachedMNode node, List<ICachedMNode> nodesToPersist) {
+    if (node.isMeasurement()) {
+      return;
+    }
     Iterator<ICachedMNode> bufferIterator =
         getCachedMNodeContainer(node).getChildrenBufferIterator();
-
     if (bufferIterator.hasNext()) {
       nodesToPersist.add(node);
-    }
-
-    ICachedMNode child;
-    while (bufferIterator.hasNext()) {
-      child = bufferIterator.next();
-      collectVolatileNodes(child, nodesToPersist);
+      ICachedMNode child;
+      while (bufferIterator.hasNext()) {
+        child = bufferIterator.next();
+        collectVolatileNodes(child, nodesToPersist);
+      }
+    } else {
+      // add back to cache
+      CacheEntry cacheEntry = getCacheEntry(node);
+      while (cacheEntry != null && !isInNodeCache(cacheEntry)) {
+        addToNodeCache(cacheEntry, node);
+        node = node.getParent();
+        cacheEntry = getCacheEntry(node);
+      }
```

Comment:
Maybe we can do this when executing deleting node.

Reply:
It's costing to judge whether a node has some child in buffer. Update the status during flushing is the fastest way.